app/vmselect/vmui.go

```diff
@@ -0,0 +1,123 @@
+package vmselect
+
+import (
+	"encoding/json"
+	"flag"
+	"fmt"
+	"net/http"
+	"os"
+	"path/filepath"
+
+	"github.com/VictoriaMetrics/VictoriaMetrics/lib/fs"
+)
+
+var (
+	vmuiCustomDashboardsPath = flag.String("vmui.customDashboardsPath", "", "Optional path to vmui predefined dashboards")
+)
+
+// dashboardSetting represents dashboard settings file struct
+type dashboardSetting struct {
+	Title    string         `json:"title,omitempty"`
+	Filename string         `json:"filename,omitempty"`
+	Rows     []dashboardRow `json:"rows"`
+}
+
+// panelSettings represents fields which used to show graph
+type panelSettings struct {
+	Title       string   `json:"title,omitempty"`
+	Description string   `json:"description,omitempty"`
+	Unit        string   `json:"unit,omitempty"`
+	Expr        []string `json:"expr"`
+	Alias       []string `json:"alias,omitempty"`
+	ShowLegend  bool     `json:"showLegend,omitempty"`
+	Width       int      `json:"width,omitempty"`
+}
+
+// dashboardRow represents panels on dashboard
+type dashboardRow struct {
+	Title  string          `json:"title,omitempty"`
+	Panels []panelSettings `json:"panels"`
+}
+
+// dashboardsData represents all dashboards settings
+type dashboardsData struct {
+	DashboardsSettings []dashboardSetting `json:"dashboardsSettings"`
+}
+
+func handleVMUICustomDashboards(w http.ResponseWriter) {
```

Comment:
Does it re-read all the custom dashboards on each request?

Reply:
@valyala 

> On the other hand, vmui should request dashboard configs from the server only once when switching to the dashboards tab.

This is contrary to https://github.com/VictoriaMetrics/VictoriaMetrics/pull/3545#discussion_r1058675084

> ...Note that by default the dashboards tab should be hidden as described at #3322 (comment).  

We can load dashboards as the application starts and store the result until the browser is reloaded.

Reply:
@valyala what about loading the dashboards on startup and then watching the file change in background with, lets say, interval of 15s? If changes are detected - reload the assets. No changes - we serve previously parsed json files.

Reply:
@valyala , @hagen1778 I have implemented a simple cache that will update dashboards settings if the new request is after 10 minutes. I think this will avoid possible many disk reads